sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/QueueClient.java

```diff
@@ -882,8 +1081,118 @@ public PagedIterable<QueueMessageItem> receiveMessages(Integer maxMessages) {
+
+    PagedIterable<QueueMessageItem> receiveMessagesWithOptionalTimeout(Integer maxMessages, Duration visibilityTimeout,
+        Duration timeout, Context context) {
+        Context finalContext = context == null ? Context.NONE : context;
+        Integer visibilityTimeoutInSeconds = (visibilityTimeout == null) ? null : (int) visibilityTimeout.getSeconds();
+        try {
+            Supplier<ResponseBase<MessagesDequeueHeaders, List<QueueMessageItemInternal>>> operation = () ->
+                this.azureQueueStorage.getMessages()
+                    .dequeueWithResponse(queueName, maxMessages, visibilityTimeoutInSeconds, null, null, finalContext);
+
+            ResponseBase<MessagesDequeueHeaders, List<QueueMessageItemInternal>> response = timeout != null
+                ? THREAD_POOL.submit(operation::get).get(timeout.toMillis(), TimeUnit.MILLISECONDS) : operation.get();
+
+            PagedResponseBase<MessagesDequeueHeaders, QueueMessageItem> transformedMessages =
+                transformMessagesDequeueResponse(response);
+
+            Supplier<PagedResponse<QueueMessageItem>> res = () -> new PagedResponseBase<>(
+                response.getRequest(),
+                response.getStatusCode(),
+                response.getHeaders(),
+                transformedMessages,
+                response.getDeserializedHeaders());
+
+            return new PagedIterable<>(res);
+
+        } catch (RuntimeException e) {
+            throw LOGGER.logExceptionAsError(e);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            throw LOGGER.logExceptionAsError(new RuntimeException(e));
+        }
+    }
+
+    private PagedResponseBase<MessagesDequeueHeaders, QueueMessageItem> transformMessagesDequeueResponse(
+        ResponseBase<MessagesDequeueHeaders, List<QueueMessageItemInternal>> response) {
+        List<QueueMessageItemInternal> queueMessageInternalItems = response.getValue();
+        if (queueMessageInternalItems == null) {
+            queueMessageInternalItems = Collections.emptyList();
+        }
+        List<QueueMessageItem> messageItems = new ArrayList<>();
+        for (QueueMessageItemInternal queueMessageInternalItem : queueMessageInternalItems) {
+            try {
+                QueueMessageItem decodedMessage =
+                    transformQueueMessageItemInternal(queueMessageInternalItem, messageEncoding);
+                messageItems.add(decodedMessage);
+
+            } catch (IllegalArgumentException e) {
+                if (processMessageDecodingErrorAsyncHandler != null) {
+                    QueueMessageItem transformedQueueMessageItem =
+                        transformQueueMessageItemInternal(queueMessageInternalItem, QueueMessageEncoding.NONE);
+
+                    processMessageDecodingErrorAsyncHandler.apply(new QueueMessageDecodingError(
+                        new QueueAsyncClient(azureQueueStorage, queueName, accountName, serviceVersion, messageEncoding,
```

Comment:
Why do we create new client each time? Can we just create this async client instance once and reuse it each time this `catch` block is executed?